internal/api/token_refresh.go

```diff
@@ -36,111 +39,152 @@ func (a *API) RefreshTokenGrant(ctx context.Context, w http.ResponseWriter, r *h
-		if terr != nil {
-			if models.IsNotFoundError(terr) {
-				return oauthError("invalid_grant", "Invalid Refresh Token: Refresh Token Not Found")
+			if !notAfter.IsZero() && time.Now().UTC().After(notAfter) {
+				return oauthError("invalid_grant", "Invalid Refresh Token: Session Expired")
 			}
-			return internalServerError(terr.Error())
 		}
 
-		// refresh token row and session are locked, cannot be concurrently refreshed
-
-		if token.Revoked {
-			a.clearCookieTokens(config, w)
-			// For a revoked refresh token to be reused, it has to fall within the reuse interval.
-
-			reuseUntil := token.UpdatedAt.Add(
-				time.Second * time.Duration(config.Security.RefreshTokenReuseInterval))
-
-			if time.Now().After(reuseUntil) {
-				// not OK to reuse this token
+		// Basic checks above passed, now we need to serialize access
+		// to the session in a transaction so that there's no
+		// concurrent modification. In the event that the refresh
+		// token's row or session is locked, the transaction is closed
+		// and the whole process will be retried a bit later so that
+		// the connection pool does not get exhausted.
+
+		var tokenString string
+		var expiresAt int64
+		var newTokenResponse *AccessTokenResponse
+
+		err = db.Transaction(func(tx *storage.Connection) error {
+			user, token, _, terr := models.FindUserWithRefreshToken(tx, params.RefreshToken, true /* forUpdate */)
+			if terr != nil {
+				if models.IsNotFoundError(terr) {
+					// because forUpdate was set, and the
+					// previous check outside the
+					// transaction found a refresh token
+					// and session, but now we're getting a
+					// IsNotFoundError, this means that the
+					// refresh token row and session are
+					// probably locked so we need to retry
+					// in a few milliseconds.
+					retry = true
+					return terr
+				}
+				return internalServerError(terr.Error())
+			}
 
-				if config.Security.RefreshTokenRotationEnabled {
-					// Revoke all tokens in token family
-					err = db.Transaction(func(tx *storage.Connection) error {
-						var terr error
-						if terr = models.RevokeTokenFamily(tx, token); terr != nil {
-							return terr
+			// refresh token row and session are locked at this
+			// point, cannot be concurrently refreshed
+
+			if token.Revoked {
+				a.clearCookieTokens(config, w)
+				// For a revoked refresh token to be reused, it
+				// has to fall within the reuse interval.
+
+				reuseUntil := token.UpdatedAt.Add(
+					time.Second * time.Duration(config.Security.RefreshTokenReuseInterval))
+
+				if time.Now().After(reuseUntil) {
+					// not OK to reuse this token
+
+					if config.Security.RefreshTokenRotationEnabled {
+						// Revoke all tokens in token family
+						err = db.Transaction(func(tx *storage.Connection) error {
+							var terr error
+							if terr = models.RevokeTokenFamily(tx, token); terr != nil {
+								return terr
+							}
+							return nil
+						})
```

Comment:
seems like this was carried over from the old implementation but why do we need to open another transaction here? can't we just use the current one?